src/searchbar.cpp

```diff
@@ -332,12 +340,13 @@ void SearchBarLineEdit::fetchSuggestions(NewSuggestionHandlerFuncPtr callback)
     const int start = m_suggestionModel.countOfRegularSuggestions();
     const auto suggestionWorker = new SuggestionListWorker(m_searchbarInput, m_token, start, this);
     connect(suggestionWorker, &SuggestionListWorker::searchFinished, this,
-            [=] (const QList<SuggestionData>& suggestionList, int token) {
+            [=] (const QList<SuggestionData>& suggestionList, const QString& text, int token) {
```

Comment:
You could have passed the search text as a captured value of the lambda. But your implementation drew my attention to the not so self-explaining `token`/`m_token` thing. Let's improve the design as follows:

1. Replace  `int SearchBarLineEdit::m_token` with `QString SearchBarLineEdit::m_suggestionWereRequestedFor` (or maybe we can instead reuse the existing `m_searchbarInput` data member?)
2. Get rid of the `token` parameter in `searchFinished()` signal of `SuggestionListWorker` - use the new `text` parameter instead (BTW, the latter better be renamed to `searchText` and moved to the beginning of the parameter list).

Reply:
@veloman-yunkan Removing `m_token` might introduce a race condition bug:
1. user type "i", and suggestions are retrived.
2. user trigger fetch more.
3. user further type to "i am", and new suggestions are triggered.
4. user immediately types back to "i", and the fetch more from 2. returns from asynchronous call back.
5. the insertion order can be raced, and m_completer.complete() is not called before fetch more.

Reply:
Good analysis! You are a shrewd antiracist! :)